pkg/azure/validation/validation_test.go

```diff
@@ -0,0 +1,477 @@
+			subnetInfo := api.AzureSubnetInfo{
+				VnetName:   entry.vnetName,
+				SubnetName: entry.subnetName,
+			}
+			errList := validateSubnetInfo(subnetInfo, fldPath)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+
+}
+
+func TestValidateHardwareProfile(t *testing.T) {
+	fldPath := field.NewPath("providerSpec.properties.hardwareProfile")
+	hwProfile := api.AzureHardwareProfile{}
+	g := NewWithT(t)
+	errList := validateHardwareProfile(hwProfile, fldPath)
+	g.Expect(len(errList)).To(Equal(1))
+	g.Expect(errList).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.properties.hardwareProfile.vmSize")}))))
+}
+
+func TestValidateOSDisk(t *testing.T) {
+	fldPath := field.NewPath("providerSpec.properties.storageProfile.osDisk")
+	table := []struct {
+		description    string
+		osDisk         api.AzureOSDisk
+		expectedErrors int
+		matcher        gomegatypes.GomegaMatcher
+	}{
+		{
+			"should forbid empty createOption",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: 20, CreateOption: ""}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.properties.storageProfile.osDisk.createOption")}))),
+		},
+		{
+			"should forbid zero osDisk size",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: 0, CreateOption: "Create"}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeInvalid), "Field": Equal("providerSpec.properties.storageProfile.osDisk.diskSizeGB")}))),
+		},
+		{
+			"should forbid negative osDisk size",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: -10, CreateOption: "Create"}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeInvalid), "Field": Equal("providerSpec.properties.storageProfile.osDisk.diskSizeGB")}))),
+		},
+	}
+
+	g := NewWithT(t)
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			errList := validateOSDisk(entry.osDisk, fldPath)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+}
+
+func TestValidateOSProfile(t *testing.T) {
+	fldPath := field.NewPath("providerSpec.properties.osProfile")
+	osProfile := api.AzureOSProfile{
+		ComputerName:  "bingo",
+		AdminUsername: "",
+		LinuxConfiguration: api.AzureLinuxConfiguration{
+			DisablePasswordAuthentication: true,
+			SSH:                           api.AzureSSHConfiguration{},
+		},
+	}
+	g := NewWithT(t)
+	errList := validateOSProfile(osProfile, fldPath)
+	g.Expect(len(errList)).To(Equal(1))
+	g.Expect(errList).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.properties.osProfile.adminUsername")}))))
+}
+
+func TestValidateDataDisks(t *testing.T) {
```

Comment:
add a test case to validate nil lun also

Reply:
added. thanks for point it out.